core/controllers/story_viewer.py

```diff
@@ -128,6 +120,69 @@ def post(self, story_id, node_id):
                     next_exp_ids = [node.exploration_id]
                     next_node_id = node.id
                     break
+        return (next_exp_ids, next_node_id, completed_node_ids)
+
+    @acl_decorators.can_access_story_viewer_page
+    def get(self, story_id, node_id):
+        """Handles GET requests."""
+        (
+            _, _, classroom_url_fragment, topic_url_fragment,
+            story_url_fragment, node_id) = self.request.path.split('/')
+        story = story_fetchers.get_story_by_id(story_id)
+        completed_nodes = story_fetchers.get_completed_nodes_in_story(
+            self.user_id, story_id)
+        ordered_nodes = [
+            node for node in story.story_contents.get_ordered_nodes()
+        ]
+
+        # In case the user is a returning user and has completed nodes in the
+        # past, redirect to the story page so that the user can continue from
+        # where they had left off.
+        # If the node id is not the first node in the story, redirect to
+        # the story page.
+        if completed_nodes or node_id != ordered_nodes[0].id:
+            self.redirect(
+                '/learn/%s/%s/story/%s' % (
+                    classroom_url_fragment, topic_url_fragment,
+                    story_url_fragment))
+            return
+
+        (next_exp_ids, next_node_id, _) = (
+            self._record_node_completion(story_id, node_id, [], ordered_nodes))
```

Comment:
In this place, the `next_node_id` can only be the first node right? Because anyone with completed nodes is redirected and if the node_id is not the first node ID then the user is also redirected.

Reply:
Ah, I get it now, do we actualy record any completion here (because the function is named `_record_node_completion`) might be better to use some different name.

Reply:
We actually do record completion of the node in the story. See this:

https://github.com/oppia/oppia/blob/fca6f514a75a5422ce7e0009f51e218dd5d57655/core/controllers/story_viewer.py#L110-L111

Reply:
Ah, I see.